src/core/stdcpp/vector.d

```diff
@@ -0,0 +1,816 @@
+/**
+ * D header file for interaction with C++ std::vector.
+ *
+ * Copyright: Copyright (c) 2018 D Language Foundation
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Guillaume Chatelet
+ *            Manu Evans
+ * Source:    $(DRUNTIMESRC core/stdcpp/vector.d)
+ */
+
+module core.stdcpp.vector;
+
+///////////////////////////////////////////////////////////////////////////////
+// std::vector declaration.
+//
+// Current caveats :
+// - missing noexcept
+// - nothrow @trusted @nogc for most functions depend on knowledge
+//   of T's construction/destruction/assignment semantics
+///////////////////////////////////////////////////////////////////////////////
+
+import core.stdcpp.allocator;
+
+enum DefaultConstruct { value }
+
+/// Constructor argument for default construction
+enum Default = DefaultConstruct();
+
+extern(C++, "std"):
+
+extern(C++, class) struct vector(T, Alloc = allocator!T)
+{
+    import core.lifetime : forward, move, moveEmplace, core_emplace = emplace;
+
+    static assert(!is(T == bool), "vector!bool not supported!");
+extern(D):
+
+    ///
+    alias size_type = size_t;
+    ///
+    alias difference_type = ptrdiff_t;
+    ///
+    alias value_type = T;
+    ///
+    alias allocator_type = Alloc;
+    ///
+    alias pointer = T*;
+    ///
+    alias const_pointer = const(T)*;
+
+    ///
+    alias as_array this;
+
+    /// MSVC allocates on default initialisation in debug, which can't be modelled by D `struct`
+    @disable this();
+
+    ///
+    alias length = size;
+    ///
+    alias opDollar = length;
+
+    ///
+    ref inout(T) front() inout pure nothrow @safe @nogc                     { return this[0]; }
+    ///
+    ref inout(T) back() inout pure nothrow @safe @nogc                      { return this[$-1]; }
+
+
+    // WIP...
```

Comment:
You going to leave this for later?

Reply:
I'll be continuing on other platforms (clang next) and fleshing out the features as I go.
Some of those are implemented for windows. I probably won't do the iterator ones; there's no reason to mirror the C++ iterators in D; we have ranges.

Reply:
But this is such a mammoth thing; it'd be much easier if I could do it a little more incrementally, and this is a good sync-point, because it's working on windows at least.
Windows users can try it out, and we can start to get bug reports.

Reply:
Also, linux is still blocked on move construction (there's an interior pointer).